### ISSUE TITLE
is it possible changing SCHAC? 

### ISSUE THREAD

**Issue by skishk** [erasmus-without-paper/ewp-specs-api-iias#111]
is it possible changing SCHAC?

**Comment by janinamincer-daszkiewicz** [comment]
Better not, because it creates problems to other nodes in the network. But it happened already in the past. Please, write email to me explaining the situation and I will start correspondence with colleagues from EUF who trace changes in SCHAC codes.

**Comment by umesh-qs** [comment]
Yes it is possible to change SCHAC. There is no such restriction.

**Comment by skishk** [comment]
> Yes it is possible to change SCHAC. There is no such restriction.

really? and how could you manage hundreds of IIA mutually approved and one of the 2 partner have changed his SCHAC code?

**Comment by umesh-qs** [comment]
@skishk your original question was about possibility of changing SCHAC code. And the answer to that is yes. But we do warn our clients on the possible impact.

We do not touch mutually approved IIA unless the partner is sending a change notification request. If we get change request, it will be considered as a new IIA that will go through the approval process.

**Comment by umesh-qs** [comment]
@skishk the reason why I said yes it so that providers do not build their system on this assumption and then complain later on. Until it is officially communicated that is it not allowed, it is and will be allowed.

**Comment by skishk** [comment]
yeah thank you :relaxed:

probably as you know this problem is in production we are still waiting for a solution from the partner so i asked if someone found a good way.

but I think the right answer is: no, we can't change SCHAC code because it is like an ID of the HEI.

so I hope people stop to change that without manage the consequence :sweat_smile:

**Comment by umesh-qs** [comment]
> yeah thank you ☺️
> 
> probably as you know this problem is in production we are still waiting for a solution from the partner so i asked if someone found a good way.
> 
> but I think the right answer is: no, we can't change SCHAC code because it is like an ID of the HEI.
> 
> so I hope people stop to change that without manage the consequence 😅

May be you mean to say logical answer at this stage. It will become right answer when it is officially approved and communicated via DG EAC.
Also after we delink hash calculation from from approval process, which is under discussion, this will become easier to handle.

**Comment by georgschermann** [comment]
schac changes are rather common, at least we dealt with around 10+ by now.
We currently automatically maintain schacs via the catalogue and PIC / Erasmus code.
Existing IIAs are untouched until changes are made **internally**.
GET requests would already respond with the new schacs.

**Comment by skishk** [comment]
> schac changes are rather common, at least we dealt with around 10+ by now.
> We currently automatically maintain schacs via the catalogue and PIC / Erasmus code.
> Existing IIAs are untouched until changes are made **internally**.
> GET requests would already respond with the new schacs.

That means condition hash is changing and the institute is changed .
So are invalid IIA, right?

**Comment by georgschermann** [comment]
yes the IIAs are invalid (or rather un-approved) at this point, each partner should have a copy of the approved (old) version in their system.

**Comment by kkaraogl** [comment]
We also try to inform them of the consequences, upon request to change SCHAC. It is happening more and more often lately (or at least this is my perception), and we should find a way to deal with it. Could dropping hashes come sooner rather than later?

**Comment by demilatof** [comment]
> schac changes are rather common, at least we dealt with around 10+ by now.

They simply change schac code or they join together so that from two different schac code they have only one?

**Comment by skishk** [comment]
> They simply change schac code or they join together so that from two different schac code they have only one?

We faced all scenarios:
A) 1 schac of one Institute changed schac because joined other institute. so other institutes chenged their schac.
B) 1 schac of one Institute changed schac on his own institute.
C) 1 schac of differente Institute changed and each institute have a new one.

the biggest problem is: after these changes id you have with them mutually approved IIA, now they are invalid or need new approval. but, if you try to do a IIA-INDEX on the new schac you don't receive anything and obviously on the old schac you can't because is no more existing... :sweat_smile: :sob: 

I'm expecting at least if some one change his schac can change at least his it in the IIA or expose new IIA (with new IIA-ID, for us is not a problem) and star a new IIA-APPROVAL flow. but this is not happening in production :sweat_smile:

**Comment by georgschermann** [comment]
these problems seem more like provider change issues and not only schac change.
If we have a schac change, we still expose all IIAs.
If we notice a partner schac change we also try to carry over all IIAs we received by this partner.
If there is a provider change, naturally the IIAs are lost if not correctly migrated, which is not always possible.

**Comment by skishk** [comment]
so it is another blocking problem in production not managed :relieved:

**Comment by demilatof** [comment]
Well, from the Network perspective, there is no problem :wink:
You have the snapshots...

From the real world perspective it is a mess

**Comment by skishk** [comment]
partners continue to change Schac code without any communication... with mutually approved IIAs... 

could we asign a ID to each Institution and then make the schac code an attribute? so if the schac code change there is no problem... the ID is still the same...

**Comment by demilatof** [comment]
> partners continue to change Schac code without any communication... with mutually approved IIAs...
> 
> could we asign a ID to each Institution and then make the schac code an attribute? so if the schac code change there is no problem... the ID is still the same...

But the SCHAC code is also the HEI_ID; should we use the new ID as HEI_ID?

**Comment by skishk** [comment]
every good solution is welcome! 
continuing like that is not possible...

**Comment by demilatof** [comment]
> every good solution is welcome! continuing like that is not possible...

Eighty years ago, the issue was raised and solved (?) by @wrygiel in this way
https://github.com/erasmus-without-paper/ewp-specs-api-discovery/issues/4

I don't see anybody populating the required fields.
Another proof that we are plenty of must that nobody observes and nobody checks:  "**Servers MUST provide all SCHAC identifiers their HEI have used in the past**"

```
<xs:enumeration value="previous-schac">
     <xs:annotation>
         <xs:documentation>
                  A previously used SCHAC identifier. Servers MUST provide all SCHAC
                  identifiers their HEI have used in the past.

                  https://github.com/erasmus-without-paper/ewp-specs-api-discovery/issues/4
          </xs:documentation>
     </xs:annotation>
</xs:enumeration>
```
https://github.com/erasmus-without-paper/ewp-specs-api-registry/blob/00e9fcb34713c211508d639ab841509131c90a2f/catalogue.xsd#L378-L387

All that said, maybe this element (previous-schac) of any help?
Because I'm afraid that it could even be a security hole...

E.g.: I receive a request by HEI ID identified by current.edu, that declares as previous-schac previous.edu
In my system, I can retrieve all the agreements linked to current.edu and previous.edu, and everything is preserved.

But what happens if the above HEI declares as previous-schac anotherHei.edu?
I provide it data about another HEI, data that the calling HEI is not entitled to have?

**Comment by skishk** [comment]
@janinamincer-daszkiewicz 
Please, could communicate to every one that if decide to change SCHAC-CODE so the HEI-ID have to communicate it previously and clean his system to work only on the new Schac? 
Until there will be a management of this kind of change of the primary key that identify an institution.

Thank you

**Comment by janinamincer-daszkiewicz** [comment]
I will add such point to the agenda for the next Infrastructure Forum.

**Comment by skishk** [comment]
> I will add such point to the agenda for the next Infrastructure Forum.

thank you very much! it's very important this point to not create problems in production as is happening with different partners... :worried:

**Comment by janinamincer-daszkiewicz** [comment]
https://esci-sd.atlassian.net/wiki/spaces/ITSC/pages/1112178747/SCHAC+code
